### PR TITLE
Load theme for raw code

### PR DESCRIPTION
**Pull request by Myriad-Dreamin**
This PR adds an option `theme` to the `raw` element.

It introduces several changes which can be argued:

#### Set default value `txt` to lang option

Set default value `txt` to lang option, otherwise Syntect does not fill any color for text in raw code, which is not good to dark themes.

image

A `txt` highlighter will simply attach a `text.plain` scope to the whole raw text:

https://github.com/trishume/syntect/blob/f0ece85c5a2dc4436442558babd9fc159dca372b/src/parsing/syntax_set.rs#L457-L467

#### <del>Always use foreground color from the theme setting<del>

<del>Typst should not override foreground color (to black color) for dark themes. It is also meaningless to detect whether a theme is in dark style. So I removed the raw element inheriting foreground color from ancestors.<del>

<del>Under draft: It is under drafting since it introduced another non-trivial option, which needs to discuss. It adds a `styling` option which is of type...</del>

### CODE REVIEW

File: NOTICE [hunk 1]
```diff
@@ -25,6 +25,35 @@ OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ================================================================================
 The MIT License applies to:
 
+* The tmTheme in `assets/files/halcyon.tmTheme`
```
Comment (laurmaedje) [inline NOTICE:28]: You can merge this entry with the other MIT License, see egg SIL License for how it's done in other places

File: crates/typst-library/src/text/raw.rs [hunk 2]
```diff
@@ -149,17 +149,55 @@ pub struct RawElem {
...
+
+    /// Load the file for syntax highlighting in the `tmTheme` file format.
+    ///
+    /// ````example
+    ///  #set raw(theme: "halcyon.tmTheme")
+    ///
+    ///  ```typ
+    ///  = Chapter 1
+    ///  #let hi = "Hello World"
+    ///  ```
```
Comment (laurmaedje) [inline crates/typst-library/src/text/raw.rs:172]:
```suggestion
    /// #set raw(theme: "halcyon.tmTheme")
    ///
    /// ```typ
    /// = Chapter 1
    /// #let hi = "Hello World"
    /// ```
```

**Comment by laurmaedje** [comment]
About not inheriting the foreground color: I see how it makes sense, but is there any way to configure it now for block-level raw text?

Regarding the `styling` function: It does not feel really Typst-y to me. This seems more like the thing a show rule should do. Which would in theory be possible with a synthesized field, but that would need careful design. I would propose leaving it out of this PR to get this merged quickly and then revisiting it.

File: NOTICE (hunk 1 above)
Comment (Myriad-Dreamin) [reply NOTICE:28]: It has the copyright `Copyright (c) 2018 Brittany Chiang`. Can I remove the copyright and merge it the other MIT License?

Comment (laurmaedje) [reply NOTICE:28]: For the SIL License, there are also different copyrights. We keep them in the list next to the affected files.

**Comment by Myriad-Dreamin** [comment]
@laurmaedje 

> About not inheriting the foreground color: I see how it makes sense, but is there any way to configure it now for block-level raw text?

A simple override rule could be implemented as follows (just what it was):

https://github.com/typst/typst/blob/1bb024ca3f75cf9cbad1ef3ff321d3c298dce609/crates/typst-library/src/text/raw.rs#L335-L342

The problem is that we cannot enable it by default, since it corrupts many themes. We could consider configure it in the `styling`. However, I'm unsure about the best design for modifying theme settings at the moment.

------------------------

> Which would in theory be possible with a synthesized field, but that would need careful design.

If we don't have a `styling` option, we cannot customize a box with the same background color as that in the theme setting. But as you said, we can leave it out of this PR for future works. Users can copy the background color from tmTheme files manually.

I wasn't previously familiar with the concept of a "synthesized field". If you could provide guidance on which files to refer to, I can check it and rework on `styling`.

**Comment by laurmaedje** [comment]
> Users can copy the background color from tmTheme files manually.

I think that's good enough for now. They'll have to write a show rule for the rectangle anyway, so copying the color won't be that big of a deal. Of course, reading it is cleaner and more composable, but it's a minor issue for now.

Synthesized fields are basically field that are constructed from other fields in `impl Synthesize for ...`. You can see how it works for `FigureElem`. That way, we could have a dictionary with relevant theme properties synthesized. But I'm not sure how we would name the synthesized field. I'm also not a big fan of the synthesized fields anyway, so I'm not sure whether I want to add even more of them (which would make it more difficult to get rid of them in the future).

**Comment by laurmaedje** [comment]
So, in principle, this looks good to merge now. The only thing I'm still unsure about is the foreground thing. Even if we use no theme at all, we couldn't affect the text color anymore because it defaults to the one from THEME, if I understand correctly.

**Comment by Myriad-Dreamin** [comment]
> we couldn't affect the text color

I understand it now. This could be weird if we cannot affect a range of text consistently. For example, ``#text(fill: red, [`red code`])``. Both of scenarios should works: we should not simply view text in raw blocks as `TextElem` since https://github.com/typst/typst/pull/1675; At the same time, we must provide a more proper way to affect them.

So we should find a way to make theme settings available in context of `show` rules, so that users can write `#show raw(it) => show_red_code(it)`. In other word, it is a better decision to postpone merge of this PR at least before v0.7.0.

------

The efforts related to `show raw` rules would come in a new PR and block this PR. For synthesized fields, we may start from supporting `it.foreground` and `it.background` for minimum viable production.

**Comment by Myriad-Dreamin** [comment]
@laurmaedje I have rolled back the changes about foreground color in the commit https://github.com/typst/typst/pull/1675/commits/42d7b9777bc02d15d6cef42dc57ab97831827853

The rest commits are kept.

**Comment by reknih** [comment]
Paint bucket icon

I have prepared an icon for `tmTheme` files in the web app

**Comment by Myriad-Dreamin** [comment]
> Paint bucket icon
> 
> I have prepared an icon for `tmTheme` files in the web app

lovely icon~

**Comment by laurmaedje** [comment]
Thank you! Also for your patience with the foreground/background discussion. I think of this as the minimal, initial approach and am optimistic that we'll find a good solution for foreground/background handling down the road.

**Pull request merged**